server/src/main/java/org/elasticsearch/common/settings/Setting.java

```diff
@@ -814,12 +823,20 @@ public String getNamespace(Setting<T> concreteSetting) {
     public interface Validator<T> {
 
         /**
-         * The validation routine for this validator.
+         * Validate this setting's value in isolation.
+         *
+         * @param value the value of this setting
+         */
+        void validate(T value);
+
+        /**
+         * Validate this setting against its dependencies, specified by {@link #settings()}.
          *
          * @param value    the value of this setting
          * @param settings a map from the settings specified by {@link #settings()}} to their values
          */
-        void validate(T value, Map<Setting<T>, T> settings);
+        default void validate(T value, Map<Setting<T>, T> settings) {
```

Comment:
does it make sense to delegate to `validate(T value)` here?

Reply:
from an interface perspective I think we might be able to only have `validate(T value, Optional<Map<Setting<T>, T>> settings)` and instead check the optional if needed?

Reply:
Thanks @s1monw.

I'd prefer to keep these two API as most clients will only be interested in validating a setting in isolation (i.e. not against dependencies).
Besides, we do a fail-fast validation [here](https://github.com/elastic/elasticsearch/pull/34184/files/e83e210c8d35d427b0f35d851e65b1169f6d2b43#diff-3b117ec54ae350c750ab41abc7e4327fR721) where we don't have resolved setting dependencies yet. Therefore, if we want to keep this fail-fast shortcut, any client will have to implement an `if optional present/absent` block even if he's interested in validating a setting against its dependencies.

What do you think about these two points?

Reply:
I think I prefer the two methods to the optional parameter too. In practice there are very few dependencies between settings, so the one-argument lambda is less noisy in almost all cases. Settings _with_ dependencies can't use a lambda anyway because they have to override the `settings()` function too.

I also think there's no need for the dependent validation to delegate to the other one, because we call both during settings validation. However I think this should be mentioned in the Javadocs as implementors shouldn't need to go to the implementation to check this (and I know I'd forget and have to check). Also the interface-level comment needs updating as it's no longer wholly accurate.

Reply:
Hi @DaveCTurner, here is a additional commit 7c33bfe6eb9cb43412bbae0a723657fd96e0a7b9 to improve the `Validator` class documentation, thank you.

Reply:
I meant _developers who will implement this interface_.

We tried to make javadoc more explicit, but I see what you mean: the API doesn't enforce the developer to ask himself whether or not he would like to implement dependencies validation. But as it's an edge case, I still hope the developer would read the javadoc or jump to the interface definition.